### DISCUSSION TITLE
Why software pipeline is needed since warp scheduler does all the pipeline work for us?

### DISCUSSION THREAD

**Discussion by wzhcz8902** [NVIDIA/cutlass#1482] [Q&A]
The moment I saw this doc https://github.com/NVIDIA/cutlass/blob/main/media/docs/pipeline.md, I just had this question, why do we need to implement a pipeline since warp scheduler does it for us in terms of keeping compute resources busy. Can you list some aspects that warp scheduler can not handle well in pipeline, resulting in the necessary of software pipeline?
Also, in the example at the end of the above document, two producer threads and one consumer thread are in the same warp or they refer to 3 threads in 3 warps?

**Comment by hwu36** [comment]
@IonThruster

**Reply by wzhcz8902** [reply]
@thakkarV Hello, can you help me figure out this question?

**Reply by thakkarV** [reply]
I'm not sure I understand your question. Software pipelining to tolerate global memory latency isn't something new. We've done two stage kernels since Volta. The only thing different with newer architectures is even deeper pipelines

**Reply by wzhcz8902** [reply]
@thakkarV Thanks for trying to help, I will look into the code to get more details.

**Comment by IonThruster** [comment]
There are a few places where the term "Pipeline" occurs in CUTLASS - all related to software techniques. One is the "Global Memory -> Shared memory software pipelining strategy" used to hide memory latency, the other is the Asynchronous Pipeline classes which help handle this software pipelining strategy (this is the link shared in the original question) and finally there are also Multi-Stage pipelined kernels which refers to how the kernel's MMA operations are pipelined. The overarching goal of all these software techniques is to maximize throughput of MMA. 

As for the warp scheduler, that is a hardware level optimization for latency hiding - one CUTLASS very much exploits as well by means of having multiple threads ready to issue loads or store or MMA operations. That said, often times - it is not sufficient to maximize throughput. Hope that helps.

**Reply by wzhcz8902** [reply]
@IonThruster Thanks for clarifying, but I just can not wrap head around it. The purpose of the software pipeline is to maximize the throughput by having all memory access and compute units work concurrently, which is just what warp scheduler does. Unless due to some reason, the warp scheduler can not handle as well as the software pipeline does.

**Reply by IonThruster** [reply]
Typically to extract parallelism / concurrency out of a GPU - there are 2 dimensions to optimize across : 
1. Maximize independent work per worker (imagine unrolled loops, deep pipelines etc.)
2. Maximize number of such workers (so that the HW warp scheduler can swap out when all independent work for a given worker has been is exhausted)

For some algorithms just #2 is enough - but for cases like MatMul where one is dealing with considerably larger latencies - it is best to employ both strategies. Also given how the MMA throughput has improved tremendously over the last several years while at the same time memory latency hasn't improved much / at all - the effective latency coverage needed to keep the MMA saturated requires one to exploit both techniques mentioned above.

**Reply by wzhcz8902** [reply]
I might get a little sense of the necessity of software pipeline, which might help a worker/warp utilize the tensor core as much as it can, without being swapped out due to absence of operands. So in other cases where tensor core is not utilized, explicit software pipeline will not be necessary, and warp scheduler will be sufficient?

**Reply by wzhcz8902** [reply]
And I think my next question is what is the guideline when to use software pipeline.

**Reply by IonThruster** [reply]
We always recommended to use some level of software pipelining for GEMMs and Convolutions in order to maximize MMA throughput. @hwu36 can correct me if otherwise.

**Reply by wzhcz8902** [reply]
Another question I have is, for tasks that are bounded by global<->shared memory accesses, deep pipelines might not be needed. For example in the cases of flash attention, these gemms are usually memory bound, may not benefit much from deep pipelines. Can you comment on that?

**Reply by IonThruster** [reply]
If you know that your algorithm is bound by memory bandwidth, then the depth of the pipeline can be tweaked just enough to maximize it. Can it be completely avoided depends on the nature of the memory subsystem capability and the number of such parallel workers in the GPU.

**Reply by wzhcz8902** [reply]
One difference I noticed between warp scheduler pipeline and software pipeline is that, the former deals with the concurrency of memory accesses and computation between different warps, while the later deals with the concurrency in the same warp. In the case where there is only one warp allocated in each of the four sections in a SM, software pipeline does help the performance by increasing the concurrency in a single warp. More SMs are added in the newer generations of GPUs, which makes the low occupancy of SMs more severe, thus software pipeline is more needed to maximize the throughput, I guess.
